torch/csrc/autograd/profiler.cpp

```diff
@@ -156,98 +281,126 @@ void enableProfiler(ProfilerConfig config) {
               inputSizes.emplace_back();
             }
           }
-          pushRange(fn.name(), msg, fn.seqNr(), std::move(inputSizes));
+          state_ptr->pushRange(fn.name(), msg, fn.seqNr(), std::move(inputSizes));
         } else {
-          pushRange(fn.name(), msg, fn.seqNr(), {});
+          state_ptr->pushRange(fn.name(), msg, fn.seqNr(), {});
         }
         return true;
       },
       [](const RecordFunction& fn) {
-        if (fn.getStartCallbacksThreadId() !=
-                RecordFunction::currentThreadId()) {
-          // If we're not in a thread that ran start callbacks, then find
-          // the eventList that was created for the original thread_id. Then,
-          // record the end event on this list so that the block is added to
-          // the correct list, instead of to a new list. This should only run
-          // when calling RecordFunction::end() in a different thread.
-          if (state == ProfilerState::Disabled) {
-            return;
-          } else {
-            std::lock_guard<std::mutex> guard(all_event_lists_map_mutex);
-            const auto& eventListIter =
-                all_event_lists_map.find(fn.getStartCallbacksThreadId());
-            TORCH_INTERNAL_ASSERT(
-                eventListIter != all_event_lists_map.end(),
-                "Did not find thread_id matching ",
-                fn.getStartCallbacksThreadId());
-
-            auto& eventList = eventListIter->second;
-            eventList->record(
-                      EventKind::PopRange,
-                      StringView(""),
-                      fn.getStartCallbacksThreadId(),
-                      state == ProfilerState::CUDA);
-          }
-        } else {
-          popRange();
+        const auto& state = at::ThreadLocalDebugInfo::get(at::DebugInfoKind::PROFILER_STATE);
+        auto state_ptr = dynamic_cast<ProfilerThreadLocalState*>(state.get());
+        if (!state_ptr || state_ptr->config().state == ProfilerState::Disabled) {
+          return;
         }
+        state_ptr->popRange(fn.getStartCallbacksThreadId());
       },
-      /* needs_inputs */ config.report_input_shapes,
-      /* sampling_prob */ 1.0,
-      /* scopes */ {RecordScope::FUNCTION, RecordScope::USER_SCOPE});
-  state = new_state;
-  c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Profiler, true);
+      needs_inputs,
+      /* scopes */ {RecordScope::FUNCTION, RecordScope::USER_SCOPE},
+      CallbackKind::PROFILER);
+}
+
+void removeProfilingCallbacks() {
+  removeCallback(CallbackKind::PROFILER);
+}
 
-  if(state == ProfilerState::CUDA) {
+bool unused_ = []() {
+  at::ThreadLocalState::registerThreadLocalSetting(
+    at::ThreadLocalSetting::PROFILER,
+    []() {
+      auto v = at::SettingValue();
+      v.value = (int64_t)(profiler_nested_depth_ > 0);
+      return v;
+    },
+    [](at::SettingValue v) {
```

Comment:
technically you're using it as enter/exit boundary, not really as "set value". If so - you should rename it because it's confusing. If it's indeed a setter, then successive calls should be idempotent, e.g. I should be able to do `set(123); set(123);` and always have the same effect as single `set(123);`.

Reply:
makes sense, instead of a single setter for a TLS setting, I will also add an option to specify a pair of setters for enter/exit events which can be called from ThreadLocalStateGuard ctor/dtor